### PR TITLE
[Attention Mask] Refactor all encoder-decoder attention mask

### PR DESCRIPTION
**Pull request by patrickvonplaten**
# What does this PR do?

This PR refactors the attention mask of all PT Seq2Seq models. While this is a nice improvement of life, it is also necessary to effectively add FA2 and SDPA to PT Seq2Seq models (without having to change 54+ files).

In a follow-up PR it'll be much easier to add FA2 to just Bart and most important Bart-like models.
The PR slightly goes against the single-file policy, but attention masks are really the same across models and there is also only so much they can be (causal, non-causal, windowed). I think it doesn't really hurt readability as the function are very clearly defined (create 4d attention mask from 2d).

For some very big exceptions (I found only one which is LED, see comment [here](https://github.com/huggingface/transformers/pull/27086#discussion_r1374403602)), we could just write part of the mask creation separately as is done. 

I could also give the mask creation functions a `_` to make it clearer that they are private mehtods in Transformers. Both keeping as is or changing is good for me.

@amyeroberts @LysandreJik @ArthurZucker this is ready for a review!

### CODE REVIEW

File: src/transformers/models/idefics/modeling_idefics.py [hunk 1]
```diff
@@ -444,39 +445,8 @@ def extra_repr(self) -> str:
...
-
-    expanded_mask = mask[:, None, None, :].expand(bsz, 1, tgt_len, src_len).to(dtype)
 
-    inverted_mask = 1.0 - expanded_mask
 
-    return inverted_mask.masked_fill(inverted_mask.to(torch.bool), torch.finfo(dtype).min)
-
-
-# this was adapted from LlamaRMSNorm
 class IdeficsRMSNorm(nn.Module):
```
Comment (patrickvonplaten) [inline src/transformers/models/idefics/modeling_idefics.py:450]:
```suggestion
# this was adapted from LlamaRMSNorm
class IdeficsRMSNorm(nn.Module):
```

File: src/transformers/models/falcon/modeling_falcon.py [hunk 2]
```diff
@@ -904,7 +743,7 @@ def forward(self, x: torch.Tensor) -> torch.Tensor:
 
 
 class FalconDecoderLayer(nn.Module):
-    def __init__(self, config: FalconConfig, attn_mask_converter=None):
+    def __init__(self, config: FalconConfig):
```
Comment (patrickvonplaten) [inline src/transformers/models/falcon/modeling_falcon.py:746]: Forgot to remove it in a previous PR

File: src/transformers/models/falcon/modeling_falcon.py [hunk 3]
```diff
@@ -1266,16 +1103,10 @@ def forward(
...
             # 4d mask is passed through the layers
-            if attention_mask is not None:
-                attention_mask = self.attn_mask_converter.to_4d(
-                    attention_mask, seq_length, key_value_length, dtype=inputs_embeds.dtype
-                )
-            else:
-                attention_mask = self.attn_mask_converter.to_causal_4d(
-                    batch_size, seq_length, key_value_length, dtype=inputs_embeds.dtype, device=inputs_embeds.device
-                )
+            attention_mask = prepare_4d_causal_attention_mask(
```
Comment (patrickvonplaten) [inline src/transformers/models/falcon/modeling_falcon.py:1107]: think that's actually cleaner after all

File: src/transformers/models/led/modeling_led.py [hunk 4]
```diff
@@ -74,22 +75,7 @@ def shift_tokens_right(input_ids: torch.Tensor, pad_token_id: int, decoder_start
...
-    mask.masked_fill_(mask_cond < (mask_cond + 1).view(mask.size(-1), 1), 0)
-    mask = mask.to(dtype)
-
-    if past_key_values_length > 0:
-        mask = torch.cat([torch.zeros(tgt_len, past_key_values_length, dtype=dtype), mask], dim=-1)
-    return mask[None, None, :, :].expand(bsz, 1, tgt_len, tgt_len + past_key_values_length)
-
-
-def _expand_mask(mask: torch.Tensor, dtype: torch.dtype, tgt_len: Optional[int] = None):
+def prepare_4d_attention_mask_inverted(mask: torch.Tensor, dtype: torch.dtype, tgt_len: Optional[int] = None):
```
Comment (patrickvonplaten) [inline src/transformers/models/led/modeling_led.py:78]: **Important** LED is the only exception that I've found for Seq2Seq. The reason is that LED needs a very unique mask that is composed of:
- local mask
- global mask
- no mask

Hence mask values can be both positive, negative and 0

File: src/transformers/models/llama/modeling_llama.py [hunk 5]
```diff
@@ -64,11 +65,11 @@ def _get_unpad_data(attention_mask):
     )
 
 
-def _expand_mask(mask: torch.Tensor, dtype: torch.dtype, tgt_len: Optional[int] = None):
+def prepare_4d_attention_mask(mask: torch.Tensor, dtype: torch.dtype, tgt_len: Optional[int] = None):
```
Comment (patrickvonplaten) [inline src/transformers/models/llama/modeling_llama.py:68]:
```suggestion
def _expand_mask(mask: torch.Tensor, dtype: torch.dtype, tgt_len: Optional[int] = None):
```

Comment (patrickvonplaten) [inline src/transformers/models/llama/modeling_llama.py:68]:
```suggestion
def _expand_mask(mask: torch.Tensor, dtype: torch.dtype, tgt_len: Optional[int] = None):
```

File: src/transformers/models/llama/modeling_llama.py [hunk 6]
```diff
@@ -64,11 +65,11 @@ def _get_unpad_data(attention_mask):
...
     warnings.warn(
-        "Calling `transformers.models.llama.modeling_llama._expand_mask` is deprecated and will be removed in v4.37. Use `transformers.models.llama.modeling_llama.AttnMaskConverter._expand_mask"
+        "Calling `transformers.models.llama.modeling_llama.prepare_4d_attention_mask` is deprecated and will be removed in v4.37. Use `transformers.models.llama.modeling_llama.AttnMaskConverter.prepare_4d_attention_mask"
```
Comment (patrickvonplaten) [inline src/transformers/models/llama/modeling_llama.py:70]:
```suggestion
        "Calling `transformers.models.llama.modeling_llama.prepare_4d_attention_mask` is deprecated and will be removed in v4.37. Use `transformers.modeling_attn_mask_utils.AttnMaskConverter.prepare_4d_attention_mask"
```

File: src/transformers/models/mask2former/modeling_mask2former.py [hunk 7]
```diff
@@ -245,21 +245,6 @@ class Mask2FormerForUniversalSegmentationOutput(ModelOutput):
     attentions: Optional[Tuple[torch.FloatTensor]] = None
 
 
-# Copied from transformers.models.detr.modeling_detr._expand_mask
-def _expand_mask(mask: torch.Tensor, dtype: torch.dtype, target_len: Optional[int] = None):
```
Comment (patrickvonplaten) [inline src/transformers/models/mask2former/modeling_mask2former.py:249]: was never used

File: src/transformers/models/maskformer/modeling_maskformer.py [hunk 8]
```diff
@@ -817,18 +803,12 @@ def forward(
             hidden_states = inputs_embeds
             input_shape = inputs_embeds.size()[:-1]
 
-        combined_attention_mask = None
```
Comment (patrickvonplaten) [inline src/transformers/models/maskformer/modeling_maskformer.py:820]: this whole code makes no sense as the combined can never not be none

File: src/transformers/models/mpt/modeling_mpt.py [hunk 9]
```diff
@@ -513,13 +454,12 @@ def forward(
...
-            attention_mask,
-            input_shape=(batch_size, seq_length),
-            past_key_values_length=past_key_values_length,
+        causal_mask = prepare_4d_causal_attention_mask(
+            attention_mask, (batch_size, seq_length), inputs_embeds, past_key_values_length
         )
+        causal_mask = causal_mask.bool()
 
-        for i, (block, layer_past) in enumerate(zip(self.blocks, past_key_values)):
+        for _, (block, layer_past) in enumerate(zip(self.blocks, past_key_values)):
```
Comment (LysandreJik) [inline src/transformers/models/mpt/modeling_mpt.py:462]: Do we even need to keep the `enumerate` in that case?

File: src/transformers/modeling_attn_mask_utils.py [hunk 10]
```diff
@@ -0,0 +1,211 @@
...
+# distributed under the License is distributed on an "AS IS" BASIS,
+# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+# See the License for the specific language governing permissions and
+# limitations under the License.
+from typing import List, Optional, Tuple, Union
+
+import torch
+
+
+class AttnMaskConverter:
```
Comment (LysandreJik) [inline src/transformers/modeling_attn_mask_utils.py:19]: Could it eventually be renamed to `AttentionMaskConverter` ? :innocent: 

(personal preference, but would rather have no abbreviations if we can have the full word without too much hassle)

**Review by LysandreJik — APPROVED** [review]
As you have mentioned in your PR description, I would personally be in favor of having `_` prefixes just to ensure they aren't leveraged by third-parties thinking they're public. They're prominently displayed in the forward pass of a significant number of models, so it's worth adding the prefix.

Very cool PR! This is much cleaner this way, IMO

**Review by amyeroberts — APPROVED** [review]
Thanks for refactoring all this code. It's so much cleaner and great to see all the repeated code being deleted 🙏 

Just a few nit comments. Only request is that a handful of our biggest models have their slow tests run as a sense check before merging.

File: src/transformers/modeling_attn_mask_utils.py (hunk 10 above)
Comment (amyeroberts) [reply src/transformers/modeling_attn_mask_utils.py:19]: +1

File: tests/test_modeling_utils.py [hunk 11]
```diff
@@ -1184,3 +1186,145 @@ def test_push_to_hub_dynamic_model(self):
         config = AutoConfig.from_pretrained(f"{USER}/test-dynamic-model", trust_remote_code=True)
         new_model = AutoModel.from_config(config, trust_remote_code=True)
         self.assertEqual(new_model.__class__.__name__, "CustomModel")
+
+
+@require_torch
+class AttentionMaskTester(unittest.TestCase):
```
Comment (ArthurZucker) [inline tests/test_modeling_utils.py:1192]: would also advocate for self.assert that we use more now 😉 but it's a nit

File: templates/adding_a_new_model/cookiecutter-template-{{cookiecutter.modelname}}/modeling_{{cookiecutter.lowercase_modelname}}.py
Comment (ArthurZucker) [inline templates/adding_a_new_model/cookiecutter-template-{{cookiecutter.modelname}}/modeling_{{cookiecutter.lowercase_modelname}}.py:1]: love this

**Review by ArthurZucker — APPROVED** [review]
Love it!

File: src/transformers/models/mpt/modeling_mpt.py (hunk 9 above)
Comment (patrickvonplaten) [inline src/transformers/models/mpt/modeling_mpt.py:462]:
```suggestion
        for block, layer_past in zip(self.blocks, past_key_values):
```

File: tests/test_modeling_utils.py (hunk 11 above)
Comment (patrickvonplaten) [reply tests/test_modeling_utils.py:1192]: Hmm, I really think `assert ...` is much better to use in tests because the error message is much cleaner. E.g. if you do:

```py
self.assertTrue(expected_ids == predicted_ids)
```

assuming each is a list, you're error message will just be "the assertion is not True"

where as doing:

```py
assert expected_ids == predicted_ids
```

gives you a much better error message. Ok to change for consistency, but I really think that just doing `assert ...` is better.

File: tests/test_modeling_utils.py [hunk 12]
```diff
@@ -1184,3 +1186,145 @@ def test_push_to_hub_dynamic_model(self):
...
+        if additional_mask is not None:
+            for bsz_idx, seq_idx in additional_mask:
+                mask_2d[bsz_idx, seq_idx] = 0
+
+        mask_4d = mask_converter.to_4d(mask_2d, query_length=q_len, key_value_length=kv_len)
+
+        assert mask_4d.shape == (bsz, 1, q_len, kv_len)
+
+        context = mask_converter.sliding_window
+        if mask_converter.is_causal and context is None:
```
[the remark this answers is not in the document]
Comment (patrickvonplaten) [reply tests/test_modeling_utils.py:1212]: Hmm fair, I'd say though for tests it's ok since only us maintainers look at them. For me, the if-else statements actually helped quite a bit to map out all the different scenarios that exist (causal mask + window, causal mask + no winow, non-causal mask + window, non-causal mask + no window)

**Comment by patrickvonplaten** [comment]
Thanks for the fast reviews everyone! Ran slow tests
```
CUDA_VISIBLE_DEVICES="0" RUN_SLOW=1 pytest ...
```

for:
- Mistral
- Falcon
- LLama
- Whisper
- Bart
- LED (exceptional model)
- Bloom (uses boolean mask)

Think this should be good enough!

File: tests/test_modeling_utils.py (hunk 11 above)
Comment (ArthurZucker) [reply tests/test_modeling_utils.py:1192]: Yeah but self.assertListEqual() should give you more info no? 
I don't mine but just want consistency!

**Comment by DavidAkinpelu** [comment]
@patrickvonplaten I know this PR has been merged, but I have a question regarding the Pytorch version of BART. The implementation assumes the decoder is always used in an autoregressive manner [Pytorch version](https://github.com/huggingface/transformers/blob/8211c59b9a8fe84d2861446b26542f89a0260e64/src/transformers/models/bart/modeling_bart.py#L998) unlike the flax [version](https://github.com/huggingface/transformers/blob/8211c59b9a8fe84d2861446b26542f89a0260e64/src/transformers/models/bart/modeling_flax_bart.py#L262) . There could be cases of the decoder being used as an "encoder" and a "cross attention". In this case, the autoregressive nature is not required. While I think the default should be the autoregressive manner,  if `is_decoder` is set to false, the non-causal masking operation should be performed.

**Comment by DavidAkinpelu** [comment]
@patrickvonplaten This paper got me thinking in that direction. [Mores+](https://arxiv.org/pdf/2205.04275.pdf)
